targetcli-fb: Add support for media change

bgly opened the pull request:
Depends on: https://github.com/open-iscsi/tcmu-runner/pull/411

Signed-off-by: Bryant G. Ly <bryantly@linux.vnet.ibm.com>

mikechristie commented:
Andy, just wanted to give you some back story on this because in a past discussion I talked about dropping dbus use.

We are trying to avoid going from rtslib -> kernel -> runner

because updating the kernel is a pain and also there are issues with detecting if daemons are up with the current multicast/ns type of call we are doing in the kernel for nl.

So we are just having rtslib go directly to the daemon.

mikechristie commented on targetcli/ui_backstore.py:
```diff
@@ -651,6 +651,12 @@ def ui_command_create(self, name, size, cfgstring, wwn=None,
+        config = self.handler + "/" + cfgstring
+
+        ok, errmsg = self.iface.MediaChange('(sts)', name, size, config)
+        if ok != 0: raise ExecutionError("MediaChange failed")
```
Why is the command not on the device, so you don't have to pass in the device name?

Do we also want to limit this command to certain types of devices like optical? Maybe that check should go in tcmu-runner though. Or, should the command be more generic like just "reconfigure" and then let tcmu-runner translate that and do the right thing.

mikechristie commented:
> But then you somehow have to detect that this generic command is a "media change"

Lol, I think I forgot that was one of the problems you were handling with this patchset. Media change seems ok to me. We can add a generic reconfigure in another PR for size and other attributes later so everything is set the same way.

I think just change the name to ChangeMedia so it matches the style of other commands like CheckConfig.

Just one other question. Where is the definition of this operation in the SCSI specs?

I found this in MMC:

4.24.2.3 Attention Conditions for Sided Discs
Devices that support changing sides shall set sense bytes SK/ASC/ASCQ to UNIT ATTENTION/NOT READY TO READY CHANGE/MEDIUM MAY HAVE CHANGED for changes that involve disc loading.

I saw in SPC there is the RMB bit in the INQUIRY data which describes if the medium is removable, and some references to removable media in SBC.

I was just trying to figure out if this should be called ChangeMedium or ChangeMedia.

bgly commented:
thats the same section I was referring too, so I guess ChangeMedium would better reflect the document.

mikechristie commented on targetcli/ui_backstore.py:
```diff
@@ -651,6 +651,12 @@ def ui_command_create(self, name, size, cfgstring, wwn=None,
                             % (name, size))
         return self.new_node(ui_so)
 
+    def ui_command_changemedium(self, name, size, cfgstring):
```
Were you going to move this to on the device so you do not have to pass in the name?

> 
> /backstores/user:file> ls
> o- user:file .............................................. [Storage Objects: 2]
>   o- lun0 ................................. [/work/luns/lun0 (1.0GiB) activated]

So instead of the backstore "user:file" it seems like it should be on the storage object lun0.

Also, do we need a try/catch somewhere? Maybe around the self.iface.ChangeMedium call. With the patch with the original naming, I killed runner while executing the command and I get this crash:

> /backstores/user:file> mediachange lun1 1G /work/luns/lun2
> Traceback (most recent call last):
>   File "/usr/bin/targetcli", line 5, in <module>
>     pkg_resources.run_script('targetcli-fb==2.1.fb48', 'targetcli')
>   File "/usr/lib/python2.7/site-packages/pkg_resources.py", line 540, in run_script
>     self.require(requires)[0].run_script(script_name, ns)
>   File "/usr/lib/python2.7/site-packages/pkg_resources.py", line 1462, in run_script
>     exec_(script_code, namespace, namespace)
>   File "/usr/lib/python2.7/site-packages/pkg_resources.py", line 41, in exec_
>     exec("""exec code in globs, locs""")
>   File "<string>", line 1, in <module>
>   File "/usr/lib/python2.7/site-packages/targetcli_fb-2.1.fb48-py2.7.egg/EGG-INFO/scripts/targetcli", line 122, in <module>
>     
>   File "/usr/lib/python2.7/site-packages/targetcli_fb-2.1.fb48-py2.7.egg/EGG-INFO/scripts/targetcli", line 112, in main
>     
>   File "/usr/lib/python2.7/site-packages/configshell_fb-1.1.24-py2.7.egg/configshell_fb/shell.py", line 905, in run_interactive
>     self._cli_loop()
>   File "/usr/lib/python2.7/site-packages/configshell_fb-1.1.24-py2.7.egg/configshell_fb/shell.py", line 734, in _cli_loop
>     self.run_cmdline(cmdline)
>   File "/usr/lib/python2.7/site-packages/configshell_fb-1.1.24-py2.7.egg/configshell_fb/shell.py", line 848, in run_cmdline
>     self._execute_command(path, command, pparams, kparams)
>   File "/usr/lib/python2.7/site-packages/configshell_fb-1.1.24-py2.7.egg/configshell_fb/shell.py", line 823, in _execute_command
>     result = target.execute_command(command, pparams, kparams)
>   File "/usr/lib/python2.7/site-packages/configshell_fb-1.1.24-py2.7.egg/configshell_fb/node.py", line 1406, in execute_command
>     return method(*pparams, **kparams)
>   File "build/bdist.linux-x86_64/egg/targetcli/ui_backstore.py", line 663, in ui_command_mediachange
>   File "/usr/lib64/python2.7/site-packages/gi/overrides/Gio.py", line 171, in __call__
>     None)
> GLib.Error: g-dbus-error-quark: GDBus.Error:org.freedesktop.DBus.Error.NoReply: Message did not receive a reply (timeout by message bus) (4)

mikechristie commented on targetcli/ui_backstore.py:
```diff
@@ -651,6 +651,12 @@ def ui_command_create(self, name, size, cfgstring, wwn=None,
+        config = self.handler + "/" + cfgstring
+
+        ok, errmsg = self.iface.ChangeMedium('(sts)', name, size, config)
+        if ok != 0: raise ExecutionError("ChangeMedium failed")
```
Could you also put the raise line on a new line so it matches the other code.

mikechristie commented on targetcli/ui_backstore.py:
```diff
@@ -785,3 +784,15 @@ def summary(self):
+
+    def ui_command_changemedium(self, size, cfgstring):
+        size = human_to_bytes(size)
+        config = self.handler + "/" + cfgstring
```
It seems to crash here. handler is set on the UIUserBackedBackstore and not set on the UIUserBackedStorageObject. I fixed up UIUserBackedStorageObject but then flip flopped and just put it on the UIUserBackedBackstore like you ad it originally since both create/remove are there.

I fixed some other items in the patch:
1. pass in device name to runner in self.iface.ChangeMedium()
2. Fix up error messages.

[targetcli-change-medium.txt](https://github.com/open-iscsi/targetcli-fb/files/2300654/targetcli-change-medium.txt)

bgly commented:
@mikechristie I changed, the patch to what you had uploaded. It seems like all you did was move the function into UIUserBackedStorageObject and fixed some error messages. Which looked good to me.

mikechristie commented:
@maurizio-lombardi 

Hey Maurizio, are you handling targetlci now? If so, the patch looks ok to me. I cc'd you guys on a mail to make sure there are no future issues with it.

@bgly, I think Maurizaio will want you to just merge the 2 patches into 1 patch. In the git commit you can just add a line that I did some work on it, so people can find me and blame if there are bugs :) You do not have to worry about anything formal for attribution. I do not care. If you need to do it for work/IBM processes just add my Signed-off-by: Mike Christie <mchristi@redhat.com>.

maurizio-lombardi commented:
@mikechristie Hi Mike, yes, I am handling it now. Thanks for reviewing the patches.
@bgly As Mike suggested, please merge the 2 patches into 1.

bgly commented:
@mikechristie @maurizio-lombardi I have squashed the commits, thanks!
